solutions/USACO open18-familytree.mdx

```diff
@@ -0,0 +1,129 @@
+---
+id: usaco-833
+source: USACO Bronze 2018 Open
+title: Family Tree
+author: Danh Ta Chi Thanh
+---
+
+## Solution
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_family_bronze_open18.html)
+
+## Alternative Solution - BFS
+
+**Time Complexity:** $\mathcal{O}(N \log N)$
+
+This problem can also be solved by using BFS to find the common ancestors and the distance between the two cows and their nearest common ancestor.
+
+<Warning>
+
+No Bronze / Silver problem has required knowledge of BFS, so feel free to skip this solution if you find it unfamiliar and come back later.
+
+</Warning>
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h> // see /general/running-code-locally
+using namespace std;
+
+//vectors
+using vi = vector<int>;
+#define all(x) begin(x), end(x)
+#define sz(x) int((x).size())
+#define pb push_back
+
+//pairs
+using pi = pair<int,int>;
+#define f first
+#define s second
+#define mp make_pair
+
+void setIO(string name = "") {
+	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
+	if (sz(name)) {
+		freopen((name+".in").c_str(), "r", stdin); // see /general/io
+		freopen((name+".out").c_str(), "w", stdout);
+	}
+}
+
+int main() {
+	setIO("family");
+	int N;
+	string c[3];
+	cin >> N >> c[1] >> c[2];
+	map<string,int> dist[3];
+	dist[1][c[1]] = 0;
+	dist[2][c[2]] = 0;
+
+	map<string,string> edge; //edge list
+
+	for (int i = 0; i < N; ++i) {
+		string u, v;
+		cin >> v >> u;
+		// If v is a mother of u, then there is an edge from node u to node v.
+		edge[u] = v;
+	}
+
+	for (int k = 1; k <= 2; ++k) {
+		queue<string> q;
+		q.push(c[k]);
+		// BFS on directed tree
+		// to find the ancestors of both cows and the distance to them
+		while (!q.empty()) {
+			string node = q.front();
+			q.pop();
+			auto E = edge.find(node);
+			if (E != edge.end()) {
+				q.push(E->s);
+				dist[k][E->s] = dist[k][node] + 1;
+			}
+		}
+	}
+
+	string ancestor;
+	int d1 = 105, d2 = 105;
+	// Finding the common ancestor with shortest distance
+	for (auto& D1 : dist[1]) {
+		auto D2 = dist[2].find(D1.f);
+		if (D2 != dist[2].end() && (d1 > D1.s || d2 > D2->s)) {

```

Comment:
This can be made linear (just by keeping distance to root and sliding them up, like a typical O(N) lca)

Reply:
How to make them linear?
I haven't known of the LCA Algorithm :(((

Reply:
Ok, so here's the general idea:

Given two nodes $a$ and $b$ are distances $d_a$ and $d_b$ away from the root (great ... great grandfather of them all), we try to find their LCA. WLOG suppose $d_a>d_b$. Slide $a$ up (move to its parent) until $d_a$ is equal to $d_b$. Then keep sliding $a$ and $b$ up until they are equal. The first number for which they are equal is the lca.

ack bad explanation but i cannot think rn sry

Reply:
You know what it's fine as is... it's N log N, and a way to find lca i've never seen before